# Data Pipelines: ETL vs ELT
Data pipeline is a generic term for moving data from one place to another. For example, it could be moving data from one server to another server.

## ETL
An [ETL pipeline](https://en.wikipedia.org/wiki/Extract,_transform,_load) is a specific kind of data pipeline and very common. ETL stands for Extract, Transform, Load. Imagine that you have a database containing web log data. Each entry contains the IP address of a user, a timestamp, and the link that the user clicked.

What if your company wanted to run an analysis of links clicked by city and by day? You would need another data set that maps IP address to a city, and you would also need to extract the day from the timestamp. With an ETL pipeline, you could run code once per day that would extract the previous day's log data, map IP address to city, aggregate link clicks by city, and then load these results into a new database. That way, a data analyst or scientist would have access to a table of log data by city and day. That is more convenient than always having to run the same complex data transformations on the raw web log data.

Before cloud computing, businesses stored their data on large, expensive, private servers. Running queries on large data sets, like raw web log data, could be expensive both economically and in terms of time. But data analysts might need to query a database multiple times even in the same day; hence, pre-aggregating the data with an ETL pipeline makes sense.

## ELT
ELT (extract, load, transform) pipelines have gained traction since the advent of cloud computing. Cloud computing has lowered the cost of storing data and running queries on large, raw data sets. Many of these cloud services, like Amazon Redshift, Google BigQuery, or IBM Db2 can be queried using SQL or a SQL-like language. With these tools, the data gets extracted, then loaded directly, and finally transformed at the end of the pipeline.

However, ETL pipelines are still used even with these cloud tools. Oftentimes, it still makes sense to run ETL pipelines and store data in a more readable or intuitive format. This can help data analysts and scientists work more efficiently as well as help an organization become more data driven.

## Outline of the Lesson
1. Extract data from different sources such as:

    - csv files
    - json files
    - APIs

2. Transform data

    - combining data from different sources
    - data cleaning
    - data types
    - parsing dates
    - file encodings
    - missing data
    - duplicate data
    - dummy variables
    - remove outliers
    - scaling features
    - engineering features

3. Load

    - send the transformed data to a database

4. ETL Pipeline

    - code an ETL pipeline

This lesson contains many Jupyter notebook exercises where you can practice the different parts of an ETL pipeline. Some of the exercises are challenging, but they also contain hints to help you get through them. You'll notice that the "transformation" section is relatively long. You'll oftentimes hear data scientists say that cleaning and transforming data is how they spend a majority of their time. This lesson reflects that reality.

## World Bank Data
In the next section, you'll find a series of exercises. These are relatively brief and focus on extracting, or in other words, reading in data from different sources. The goal is to familiarize yourself with different types of files and see how the same data can be formatted in different ways. This lesson assumes you have experience with pandas and basic programming skills.

This lesson uses data from the World Bank. The data comes from two sources:

1. World Bank Indicator Data - This data contains socio-economic indicators for countries around the world. A few example indicators include population, arable land, and central government debt.
2. World Bank Project Data - This data set contains information about World Bank project lending since 1947.

## How to Tackle the Exercises
This course assumes you have experience manipulating data with the Pandas library, which is covered in the data analyst nanodegree. Some of these transformation exercises are challenging. The most challenging exercises are marked (challenging). If an exercise is marked as a challenge, it means you’ll get something out of solving it, but it’s not essential for understanding the lesson material or for getting through the final project at the end of this data engineering course.

Throughout the exercises, you might have to read the pandas documentation or search outside the classroom for how to do a certain processing technique. That is not just expected but also encouraged. As a data scientist professional, you will oftentimes have to research how to do something on your own much like what software engineers do. See this answer on Quora about [how often do people use stackoverflow when working on data science projects?](https://www.quora.com/How-often-do-people-use-stackoverflow-when-working-on-data-science-projects).

Use Google and other search engines when you're not sure how to do something!

## What You Will do in the Next Section
In the next section of the lesson, you'll learn about the extract portion of an ETL pipeline. You’ll get practice with a series of exercises. These exercises are relatively brief and focus on extracting, or in other words, reading in data from different sources. The goal is to familiarize yourself with different types of files and see how the same data can be formatted in different ways.

For a review of pandas, click on the "Extracurricular" section of the classroom. Open the Prerequisite: Python for Data Analysis course, and go to Lesson 7: Pandas.

## Summary of the data file types you'll work with
#### CSV files
CSV stands for comma-separated values. These types of files separate values with a comma, and each entry is on a separate line. Oftentimes, the first entry will contain variable names. Here is an example of what CSV data looks like. This is an abbreviated version of the first three lines in the World Bank projects data csv file.
#### JSON
JSON is a file format with key/value pairs. It looks like a Python dictionary. The exact same CSV file represented in JSON could look like this:
#### XML
Another data format is called XML (Extensible Markup Language). XML is very similar to HTML at least in terms of formatting. The main difference between the two is that HTML has pre-defined tags that are standardized. In XML, tags can be tailored to the data set. Here is what this same data would look like as XML.
XML is falling out of favor especially because JSON tends to be easier to navigate; however, you still might come across XML data. The World Bank API, for example, can return either XML data or JSON data. From a data perspective, the process for handling HTML and XML data is essentially the same.

#### SQL databases
SQL databases store data in tables using primary and foreign keys. In a SQL database, the same data would look like this:
#### Text Files
This course won't go into much detail about text data. There are other Udacity courses, namely on natural language processing, that go into the details of processing text for machine learning.

## Extracting Data from the Web
In this lesson, you'll see how to extract data from the web using an APIs (Application Programming Interface). APIs generally provide data in either JSON or XML format.

Companies and organizations provide APIs so that programmers can access data in an official, safe way. APIs allow you to download, and sometimes even upload or modify, data from a web server without giving you direct access.

# Goal of the ETL Lesson

The main goal of this ETL pipelines lesson is to take the [World Bank Project data set](https://datacatalog.worldbank.org/dataset/world-bank-projects-operations) and merge this data with the [World Bank indicator data](https://data.worldbank.org/indicator/SP.POP.TOTL). Then you'll load the merged data into a database.

In the process, you'll need to transform these data sets in different ways. And finally, you'll code an ETL pipeline to extract, transform, and load the data all in one step.

# Extracting data from a csv file

The first step in an ETL pipeline is extraction. Data comes in all types of different formats, and you'll practice extracting data from csv files, JSON files, XML files, SQL databases, and the web.

In this first exercise, you'll practice extracting data from a CSV file and then navigating through the results. You'll see that extracting data is not always a straight-forward process.

This exercise contains a series of coding questions for you to solve. If you get stuck, there is a solution file called 1_csv_exercise_solution.ipynb. You can find this solution file by going to File->Open and then clicking on the file name.

# Part 1 projects_data.csv

You'll be using the following csv files:
* projects_data.csv
* population_data.csv

As a first step, try importing the projects data using the pandas [read_csv method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html). The file path is just '../data/projects_data.csv'. You can see the file if you click on File->Open in the workspace and open the data folder.

In [1]:
# TODO: import the projects_data.csv file using the pandas library
# Store the results in the df_projects variable
import pandas as pd
df_projects = pd.read_csv('projects_data.csv')

/Users/xuhao3/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3063: DtypeWarning: Columns (44) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Did you get a DType warning? Read about what this warning is in the [pandas documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.errors.DtypeWarning.html).

Pandas tries to figure out programatically the data type of each column (integer, float, boolean, string). In this case, pandas could not automatically figure out the data type. That is because some columns have more than one possible data types. In other words, this data is messy.

You can use the dtype option to specify the data type of each column. Because there are so many columns in this data set, you can set all columns to be strings at least for now.

Try reading in the data set again using the read_csv() method. This time, also use the option dtype=str so that pandas treats everything like a string.

In [2]:
# TODO: Read in the projects_data.csv file using the read_csv method 
# and dtype = str option
df_projects = pd.read_csv('projects_data.csv', dtype = str)

In [5]:
# Run the cell below to see what the data looks like
df_projects.head()

id  regionname                                        countryname  \
0  P162228       Other                                        World;World   
1  P163962      Africa  Democratic Republic of the Congo;Democratic Re...   
2  P167672  South Asia  People's Republic of Bangladesh;People's Repub...   
3  P158768  South Asia  Islamic  Republic of Afghanistan;Islamic  Repu...   
4  P161364      Africa  Federal Republic of Nigeria;Federal Republic o...   

  prodline                  lendinginstr lendinginstrtype  \
0       RE  Investment Project Financing               IN   
1       PE  Investment Project Financing               IN   
2       PE  Investment Project Financing               IN   
3       PE  Investment Project Financing               IN   
4       PE  Investment Project Financing               IN   

  envassesmentcategorycode supplementprojectflg productlinetype  \
0                        C                    N               L   
1                        B                    N               L   
2                      NaN                    Y               L   
3                        A                    N               L   
4                        B                    N               L   

  projectstatusdisplay  ... mjtheme3name mjtheme4name mjtheme5name  \
0               Active  ...          NaN          NaN          NaN   
1               Active  ...          NaN          NaN          NaN   
2               Active  ...          NaN          NaN          NaN   
3               Active  ...          NaN          NaN          NaN   
4               Active  ...          NaN          NaN          NaN   

                                            location  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  0002327546!$!Ogun State!$!7!$!3.58333!$!NG;000...   

                                            GeoLocID  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  0002327546;0002328925;0002565340;0002565343;00...   

                                          GeoLocName                 Latitude  \
0                                                NaN                      NaN   
1                                                NaN                      NaN   
2                                                NaN                      NaN   
3                                                NaN                      NaN   
4  Ogun State;Niger State;Abia State;Edo;Kebbi St...  7;10;5.41667;6.5;11.5;8   

                Longitude            Country Unnamed: 56  
0                     NaN                NaN         NaN  
1                     NaN                NaN         NaN  
2                     NaN                NaN         NaN  
3                     NaN                NaN         NaN  
4  3.58333;6;7.5;6;4;10.5  NG;NG;NG;NG;NG;NG         NaN  

[5 rows x 57 columns]

In [7]:
# TODO: count the number of null values in the data set
# HINT: use the isnull() and sum() methods
df_projects.isnull().sum()

id                              0
regionname                      0
countryname                     0
prodline                        0
lendinginstr                  246
lendinginstrtype              246
envassesmentcategorycode     5811
supplementprojectflg           53
productlinetype                 0
projectstatusdisplay            4
status                          4
project_name                    0
boardapprovaldate            1504
board_approval_month         1504
closingdate                  3349
lendprojectcost               125
ibrdcommamt                     0
idacommamt                      0
totalamt                        0
grantamt                        0
borrower                     5919
impagency                    6097
url                             0
projectdoc                  18248
majorsector_percent         18248
sector1                         0
sector2                      8721
sector3                     11761
sector4                     13872
sector5       

Notice that the number 18248 shows up multiple times. There is also a countryname column with 0 missing values and a Country column with 14045 missing values. This data set clearly has some issues that will need to be solved in the transform part of the pipeline.

Next, output the shape of the data frame

In [9]:
# TODO: output the shape of the data frame
df_projects.shape

(18248, 57)

There are 18248 rows in this data set. Considering many columns had 18248 NaN values, many columns in the data set are filled completely with NaN values.

# Part 2 population_data.csv
Next, use the pandas read_csv method to read in the population_data.csv file. The path to this file is "../data/population_data.csv". When you try to read in this data set using pandas, you'll get an error because there is something wrong with the data.

In [11]:
# TODO: read in the population_data.csv file using the read_csv() method
# Put the results in a variable called df_population
df_population = pd.read_csv('population_data.csv')

ParserError: Error tokenizing data. C error: Expected 3 fields in line 5, saw 63


There is something wrong with this data set. You should see an error that says "expected 3 fields in line 5, saw 63". What might have happened? Try printing out the first few lines of the data file to see what the issue might be.

In [12]:
# TODO: Print out the first 10 lines of the data set, line by line.
# HINT: You can't use the read_csv method from pandas
# HINT: to do this manually, you'll need to use pure Python
# HINT: the open(), readline(), and close() methods should be helpful
# HINT: Use a for loop
with open('population_data.csv') as f:
    for i in range(10):
        print(f.readline())

"Data Source","World Development Indicators",



"Last Updated Date","2018-06-28",



"Country Name","Country Code","Indicator Name","Indicator Code","1960","1961","1962","1963","1964","1965","1966","1967","1968","1969","1970","1971","1972","1973","1974","1975","1976","1977","1978","1979","1980","1981","1982","1983","1984","1985","1986","1987","1988","1989","1990","1991","1992","1993","1994","1995","1996","1997","1998","1999","2000","2001","2002","2003","2004","2005","2006","2007","2008","2009","2010","2011","2012","2013","2014","2015","2016","2017",

"Aruba","ABW","Population, total","SP.POP.TOTL","54211","55438","56225","56695","57032","57360","57715","58055","58386","58726","59063","59440","59840","60243","60528","60657","60586","60366","60103","59980","60096","60567","61345","62201","62836","63026","62644","61833","61079","61032","62149","64622","68235","72504","76700","80324","83200","85451","87277","89005","90853","92898","94992","97017","98737","100031","100832","101220","10135

The first four lines in the file are not properly formatted and don't contain data. Next, read in the data using the read_csv method. But this time, use the skiprows option.

In [19]:
# TODO: read in population data skipping the first four rows
# Put the results in a variable called df_population

df_population = pd.read_csv('population_data.csv', header = 2)

In [20]:
df_population.head()

Country Name Country Code     Indicator Name Indicator Code       1960  \
0        Aruba          ABW  Population, total    SP.POP.TOTL    54211.0   
1  Afghanistan          AFG  Population, total    SP.POP.TOTL  8996351.0   
2       Angola          AGO  Population, total    SP.POP.TOTL  5643182.0   
3      Albania          ALB  Population, total    SP.POP.TOTL  1608800.0   
4      Andorra          AND  Population, total    SP.POP.TOTL    13411.0   

        1961       1962       1963       1964       1965        1966  \
0    55438.0    56225.0    56695.0    57032.0    57360.0     57715.0   
1  9166764.0  9345868.0  9533954.0  9731361.0  9938414.0  10152331.0   
2  5753024.0  5866061.0  5980417.0  6093321.0  6203299.0   6309770.0   
3  1659800.0  1711319.0  1762621.0  1814135.0  1864791.0   1914573.0   
4    14375.0    15370.0    16412.0    17469.0    18549.0     19647.0   

         1967        1968        1969        1970        1971        1972  \
0     58055.0     58386.0     58726.0     59063.0     59440.0     59840.0   
1  10372630.0  10604346.0  10854428.0  11126123.0  11417825.0  11721940.0   
2   6414995.0   6523791.0   6642632.0   6776381.0   6927269.0   7094834.0   
3   1965598.0   2022272.0   2081695.0   2135479.0   2187853.0   2243126.0   
4     20758.0     21890.0     23058.0     24276.0     25559.0     26892.0   

         1973        1974        1975        1976        1977        1978  \
0     60243.0     60528.0     60657.0     60586.0     60366.0     60103.0   
1  12027822.0  12321541.0  12590286.0  12840299.0  13067538.0  13237734.0   
2   7277960.0   7474338.0   7682479.0   7900997.0   8130988.0   8376147.0   
3   2296752.0   2350124.0   2404831.0   2458526.0   2513546.0   2566266.0   
4     28232.0     29520.0     30705.0     31777.0     32771.0     33737.0   

         1979        1980        1981        1982        1983        1984  \
0     59980.0     60096.0     60567.0     61345.0     62201.0     62836.0   
1  13306695.0  13248370.0  13053954.0  12749645.0  12389269.0  12047115.0   
2   8641521.0   8929900.0   9244507.0   9582156.0   9931562.0  10277321.0   
3   2617832.0   2671997.0   2726056.0   2784278.0   2843960.0   2904429.0   
4     34818.0     36067.0     37500.0     39114.0     40867.0     42706.0   

         1985        1986        1987        1988        1989        1990  \
0     63026.0     62644.0     61833.0     61079.0     61032.0     62149.0   
1  11783050.0  11601041.0  11502761.0  11540888.0  11777609.0  12249114.0   
2  10609042.0  10921037.0  11218268.0  11513968.0  11827237.0  12171441.0   
3   2964762.0   3022635.0   3083605.0   3142336.0   3227943.0   3286542.0   
4     44600.0     46517.0     48455.0     50434.0     52448.0     54509.0   

         1991        1992        1993        1994        1995        1996  \
0     64622.0     68235.0     72504.0     76700.0     80324.0     83200.0   
1  12993657.0  13981231.0  15095099.0  16172719.0  17099541.0  17822884.0   
2  12553446.0  12968345.0  13403734.0  13841301.0  14268994.0  14682284.0   
3   3266790.0   3247039.0   3227287.0   3207536.0   3187784.0   3168033.0   
4     56671.0     58888.0     60971.0     62677.0     63850.0     64360.0   

         1997        1998        1999        2000        2001        2002  \
0     85451.0     87277.0     89005.0     90853.0     92898.0     94992.0   
1  18381605.0  18863999.0  19403676.0  20093756.0  20966463.0  21979923.0   
2  15088981.0  15504318.0  15949766.0  16440924.0  16983266.0  17572649.0   
3   3148281.0   3128530.0   3108778.0   3089027.0   3060173.0   3051010.0   
4     64327.0     64142.0     64370.0     65390.0     67341.0     70049.0   

         2003        2004        2005        2006        2007        2008  \
0     97017.0     98737.0    100031.0    100832.0    101220.0    101353.0   
1  23064851.0  24118979.0  25070798.0  25893450.0  26616792.0  27294031.0   
2  18203369.0  18865716.0  19552542.0  20262399.0  20997687.0  21759420.0   
3   3039616.0   3026939.0 

Make sure to scroll over to see what the last column looks like. That last column, called 'Unnamed: 62', doesn't look very useful and is filled with NaN values.

In [21]:
pd.set_option("display.max_rows", None)
# TODO: Count the number of null values in each column
df_population.isnull().sum()

Country Name        0
Country Code        0
Indicator Name      0
Indicator Code      0
1960                4
1961                4
1962                4
1963                4
1964                4
1965                4
1966                4
1967                4
1968                4
1969                4
1970                4
1971                4
1972                4
1973                4
1974                4
1975                4
1976                4
1977                4
1978                4
1979                4
1980                4
1981                4
1982                4
1983                4
1984                4
1985                4
1986                4
1987                4
1988                4
1989                4
1990                2
1991                2
1992                3
1993                3
1994                3
1995                2
1996                2
1997                2
1998                1
1999                1
2000                1
2001      

It looks like every year column has at least one NaN value. 

In [22]:
# TODO: Count the number of null values in each row
# HINT: In the sum method, use axis=1
df_population.isnull().sum(axis = 1)

0       1
1       1
2       1
3       1
4       1
5       1
6       1
7       1
8       1
9       1
10      1
11      1
12      1
13      1
14      1
15      1
16      1
17      1
18      1
19      1
20      1
21      1
22      1
23      1
24      1
25      1
26      1
27      1
28      1
29      1
30      1
31      1
32      1
33      1
34      1
35      1
36      1
37      1
38      1
39      1
40      1
41      1
42      1
43      1
44      1
45      1
46      1
47      1
48      1
49      1
50      1
51      1
52      1
53      1
54      1
55      1
56      1
57      1
58      1
59      1
60      1
61      1
62      1
63      1
64      1
65      1
66      1
67      7
68      1
69      1
70      1
71      1
72      1
73      1
74      1
75      1
76      1
77      1
78      1
79      1
80      1
81      1
82      1
83      1
84      1
85      1
86      1
87      1
88      1
89      1
90      1
91      1
92      1
93      1
94      1
95      1
96      1
97      1
98      1
99      1


And it looks like almost every row has only one null value. That is probably from the 'Unnamed: 62' column that doesn't have any relevant information in it. Next, drop the 'Unnamed: 62' column from the data frame.

In [23]:
# TODO: drop the 'Unnamed: 62' column from the data frame, 
# and save the results in the df_population variable

df_population.drop(columns = {'Unnamed: 62'}, inplace = True)

In [24]:
df_population.head()

Country Name Country Code     Indicator Name Indicator Code       1960  \
0        Aruba          ABW  Population, total    SP.POP.TOTL    54211.0   
1  Afghanistan          AFG  Population, total    SP.POP.TOTL  8996351.0   
2       Angola          AGO  Population, total    SP.POP.TOTL  5643182.0   
3      Albania          ALB  Population, total    SP.POP.TOTL  1608800.0   
4      Andorra          AND  Population, total    SP.POP.TOTL    13411.0   

        1961       1962       1963       1964       1965        1966  \
0    55438.0    56225.0    56695.0    57032.0    57360.0     57715.0   
1  9166764.0  9345868.0  9533954.0  9731361.0  9938414.0  10152331.0   
2  5753024.0  5866061.0  5980417.0  6093321.0  6203299.0   6309770.0   
3  1659800.0  1711319.0  1762621.0  1814135.0  1864791.0   1914573.0   
4    14375.0    15370.0    16412.0    17469.0    18549.0     19647.0   

         1967        1968        1969        1970        1971        1972  \
0     58055.0     58386.0     58726.0     59063.0     59440.0     59840.0   
1  10372630.0  10604346.0  10854428.0  11126123.0  11417825.0  11721940.0   
2   6414995.0   6523791.0   6642632.0   6776381.0   6927269.0   7094834.0   
3   1965598.0   2022272.0   2081695.0   2135479.0   2187853.0   2243126.0   
4     20758.0     21890.0     23058.0     24276.0     25559.0     26892.0   

         1973        1974        1975        1976        1977        1978  \
0     60243.0     60528.0     60657.0     60586.0     60366.0     60103.0   
1  12027822.0  12321541.0  12590286.0  12840299.0  13067538.0  13237734.0   
2   7277960.0   7474338.0   7682479.0   7900997.0   8130988.0   8376147.0   
3   2296752.0   2350124.0   2404831.0   2458526.0   2513546.0   2566266.0   
4     28232.0     29520.0     30705.0     31777.0     32771.0     33737.0   

         1979        1980        1981        1982        1983        1984  \
0     59980.0     60096.0     60567.0     61345.0     62201.0     62836.0   
1  13306695.0  13248370.0  13053954.0  12749645.0  12389269.0  12047115.0   
2   8641521.0   8929900.0   9244507.0   9582156.0   9931562.0  10277321.0   
3   2617832.0   2671997.0   2726056.0   2784278.0   2843960.0   2904429.0   
4     34818.0     36067.0     37500.0     39114.0     40867.0     42706.0   

         1985        1986        1987        1988        1989        1990  \
0     63026.0     62644.0     61833.0     61079.0     61032.0     62149.0   
1  11783050.0  11601041.0  11502761.0  11540888.0  11777609.0  12249114.0   
2  10609042.0  10921037.0  11218268.0  11513968.0  11827237.0  12171441.0   
3   2964762.0   3022635.0   3083605.0   3142336.0   3227943.0   3286542.0   
4     44600.0     46517.0     48455.0     50434.0     52448.0     54509.0   

         1991        1992        1993        1994        1995        1996  \
0     64622.0     68235.0     72504.0     76700.0     80324.0     83200.0   
1  12993657.0  13981231.0  15095099.0  16172719.0  17099541.0  17822884.0   
2  12553446.0  12968345.0  13403734.0  13841301.0  14268994.0  14682284.0   
3   3266790.0   3247039.0   3227287.0   3207536.0   3187784.0   3168033.0   
4     56671.0     58888.0     60971.0     62677.0     63850.0     64360.0   

         1997        1998        1999        2000        2001        2002  \
0     85451.0     87277.0     89005.0     90853.0     92898.0     94992.0   
1  18381605.0  18863999.0  19403676.0  20093756.0  20966463.0  21979923.0   
2  15088981.0  15504318.0  15949766.0  16440924.0  16983266.0  17572649.0   
3   3148281.0   3128530.0   3108778.0   3089027.0   3060173.0   3051010.0   
4     64327.0     64142.0     64370.0     65390.0     67341.0     70049.0   

         2003        2004        2005        2006        2007        2008  \
0     97017.0     98737.0    100031.0    100832.0    101220.0    101353.0   
1  23064851.0  24118979.0  25070798.0  25893450.0  26616792.0  27294031.0   
2  18203369.0  18865716.0  19552542.0  20262399.0  20997687.0  21759420.0   
3   3039616.0   3026939.0 

In [25]:
# Run this code cell.  
# This code outputs any row that contains a null value
# The purpose is to see what rows contain null values now that 
#   'Unnamed: 62' was dropped from the data.
df_population[df_population.isnull().any(axis=1)]

Country Name Country Code     Indicator Name Indicator Code  \
67                     Eritrea          ERI  Population, total    SP.POP.TOTL   
108             Not classified          INX  Population, total    SP.POP.TOTL   
125                     Kuwait          KWT  Population, total    SP.POP.TOTL   
194         West Bank and Gaza          PSE  Population, total    SP.POP.TOTL   
212                     Serbia          SRB  Population, total    SP.POP.TOTL   
223  Sint Maarten (Dutch part)          SXM  Population, total    SP.POP.TOTL   

          1960       1961       1962       1963       1964       1965  \
67   1397491.0  1432640.0  1469645.0  1508273.0  1548187.0  1589179.0   
108        NaN        NaN        NaN        NaN        NaN        NaN   
125   269618.0   301336.0   338296.0   379891.0   425235.0   473554.0   
194        NaN        NaN        NaN        NaN        NaN        NaN   
212        NaN        NaN        NaN        NaN        NaN        NaN   
223        NaN        NaN        NaN        NaN        NaN        NaN   

          1966       1967       1968       1969       1970       1971  \
67   1631147.0  1674204.0  1718525.0  1764343.0  1811878.0  1861199.0   
108        NaN        NaN        NaN        NaN        NaN        NaN   
125   524856.0   579007.0   634897.0   691129.0   746767.0   801142.0   
194        NaN        NaN        NaN        NaN        NaN        NaN   
212        NaN        NaN        NaN        NaN        NaN        NaN   
223        NaN        NaN        NaN        NaN        NaN        NaN   

          1972       1973       1974       1975       1976       1977  \
67   1912302.0  1965160.0  2019717.0  2075965.0  2133723.0  2193068.0   
108        NaN        NaN        NaN        NaN        NaN        NaN   
125   854604.0   908520.0   964834.0  1024940.0  1089209.0  1157033.0   
194        NaN        NaN        NaN        NaN        NaN        NaN   
212        NaN        NaN        NaN        NaN        NaN        NaN   
223        NaN        NaN        NaN        NaN        NaN        NaN   

          1978       1979       1980       1981       1982       1983  \
67   2254450.0  2318495.0  2385540.0  2454766.0  2525521.0  2598410.0   
108        NaN        NaN        NaN        NaN        NaN        NaN   
125  1227601.0  1299683.0  1372318.0  1442991.0  1511314.0  1580638.0   
194        NaN        NaN        NaN        NaN        NaN        NaN   
212        NaN        NaN        NaN        NaN        NaN        NaN   
223        NaN        NaN        NaN        NaN        NaN        NaN   

          1984       1985       1986       1987       1988       1989  \
67   2674289.0  2753151.0  2837111.0  2924349.0  3006361.0  3071771.0   
108        NaN        NaN        NaN        NaN        NaN        NaN   
125  1655833.0  1738994.0  1836105.0  1942810.0  2038885.0  2096932.0   
194        NaN        NaN        NaN        NaN        NaN        NaN   
212        NaN        NaN        NaN        NaN        NaN        NaN   
223        NaN        NaN        NaN        NaN        NaN        NaN   

          1990       1991       1992       1993       1994       1995  \
67   3113311.0  3127297.0  3118582.0  3099047.0  3085443.0  3090159.0   
108        NaN        NaN        NaN        NaN        NaN        NaN   
125  2099615.0  2035661.0        NaN        NaN        NaN  1610651.0   
194  1978248.0  2068845.0  2163591.0  2262676.0  2366298.0  2474666.0   
212  7586000.0  7595636.0  7646424.0  7699307.0  7734639.0  7625357.0   
223        NaN        NaN        NaN        NaN        NaN        NaN   

          1996       1997       1998       1999       2000       2001  \
67   3116379.0  3161350.0  3224223.0  3302263.0  3392801.0  3497124.0   
108        NaN        NaN        NaN        NaN        NaN        NaN   
125  1631740.0  1715314.0  1836353.0  1957066.0  2050741.0  2109355.0   
194  2587997.0  2706518.0  2776568.0  2848431.0  2922153.0  2997784.0   
212  7617794.

# Conclusion 

This population data doesn't look too bad. Only six rows have missing values. In the transformation part of the lesson, you'll have to deal with these missing values somehow.

If you would like to see the solution file for this exercise, go to File->Open and click on 1_csv_exercise_solution.ipynb.

In the next exercise, you'll practice extracting data json and xml files.

# Extract from JSON and XML

You'll now get practice extracting data from JSON and XML. You'll extract the same population data from the previous exercise, except the data will be in a different format.

Both JSON and XML are common formats for storing data. XML was established before JSON, and JSON has become more popular over time. They both tend to be used for sending data via web APIs, which you'll learn about later in the lesson.

Sometimes, you can obtain the same data in either JSON or XML format. The World Bank indicator data is available in either form. In this exercise, you'll use the same data except one file is formatted as JSON and the other as XML.

There is a solution file for these exercises. Go to File->Open and click on 2_extract_exercise_solution.ipynb.

# Extract JSON and JSON Exercise

First, you'll practice extracting data from a JSON file. Run the cell below to print out the first line of the JSON file.

In [26]:
### 
#   Run the following cell.
#   This cell loads a function that prints the first n lines of
#   a file.
#
#   Then this function is called on the JSON file to print out
#   the first line of the population_data.json file
###

def print_lines(n, file_name):
    f = open(file_name)
    for i in range(n):
        print(f.readline())
    f.close()
    
print_lines(1, 'population_data.json')

[{"Country Name":"Aruba","Country Code":"ABW","Indicator Name":"Population, total","Indicator Code":"SP.POP.TOTL","1960":54211.0,"1961":55438.0,"1962":56225.0,"1963":56695.0,"1964":57032.0,"1965":57360.0,"1966":57715.0,"1967":58055.0,"1968":58386.0,"1969":58726.0,"1970":59063.0,"1971":59440.0,"1972":59840.0,"1973":60243.0,"1974":60528.0,"1975":60657.0,"1976":60586.0,"1977":60366.0,"1978":60103.0,"1979":59980.0,"1980":60096.0,"1981":60567.0,"1982":61345.0,"1983":62201.0,"1984":62836.0,"1985":63026.0,"1986":62644.0,"1987":61833.0,"1988":61079.0,"1989":61032.0,"1990":62149.0,"1991":64622.0,"1992":68235.0,"1993":72504.0,"1994":76700.0,"1995":80324.0,"1996":83200.0,"1997":85451.0,"1998":87277.0,"1999":89005.0,"2000":90853.0,"2001":92898.0,"2002":94992.0,"2003":97017.0,"2004":98737.0,"2005":100031.0,"2006":100832.0,"2007":101220.0,"2008":101353.0,"2009":101453.0,"2010":101669.0,"2011":102053.0,"2012":102577.0,"2013":103187.0,"2014":103795.0,"2015":104341.0,"2016":104822.0,"2017":105264.0},{"

In [35]:
import json
with open('population_data.json','r') as jsfile:
    lines = jsfile.readlines()
    js_lines = json.loads(lines[0])

print(json.dumps(js_line[:2],indent=4)) 


[
    {
        "Country Name": "Aruba",
        "Country Code": "ABW",
        "Indicator Name": "Population, total",
        "Indicator Code": "SP.POP.TOTL",
        "1960": 54211.0,
        "1961": 55438.0,
        "1962": 56225.0,
        "1963": 56695.0,
        "1964": 57032.0,
        "1965": 57360.0,
        "1966": 57715.0,
        "1967": 58055.0,
        "1968": 58386.0,
        "1969": 58726.0,
        "1970": 59063.0,
        "1971": 59440.0,
        "1972": 59840.0,
        "1973": 60243.0,
        "1974": 60528.0,
        "1975": 60657.0,
        "1976": 60586.0,
        "1977": 60366.0,
        "1978": 60103.0,
        "1979": 59980.0,
        "1980": 60096.0,
        "1981": 60567.0,
        "1982": 61345.0,
        "1983": 62201.0,
        "1984": 62836.0,
        "1985": 63026.0,
        "1986": 62644.0,
        "1987": 61833.0,
        "1988": 61079.0,
        "1989": 61032.0,
        "1990": 62149.0,
        "1991": 64622.0,
        "1992": 68235.0,
        "1993":

The first "line" in the file is actually the entire file. JSON is a compact way of representing data in a dictionary-like format. Luckily, pandas has a method to [read in a json file](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_json.html) and parse the results for you. 

If you open the link with the documentation, you'll see there is an *orient* option that can handle JSON formatted in different ways:
```
'split' : dict like {index -> [index], columns -> [columns], data -> [values]}

'records' : list like [{column -> value}, ... , {column -> value}]

'index' : dict like {index -> {column -> value}}

'columns' : dict like {column -> {index -> value}}

'values' : just the values array
```

In this case, the JSON is formatted with a 'records' orientation, so you'll need to use that value in the read_json() method. You can tell that the format is 'records' by comparing the pattern in the documentation with the pattern in the JSON file.

Next, read in the population_data.json file using pandas.

In [30]:
# TODO: Read in the population_data.json file using pandas's 
# read_json method. Don't forget to specific the orient option
# store the results in df_json

import pandas as pd
df_json = pd.read_json('population_data.json',orient = 'records')

In [31]:
# TODO: Use the head method to see the first few rows of the resulting
df_json.head()

Country Name Country Code     Indicator Name Indicator Code       1960  \
0        Aruba          ABW  Population, total    SP.POP.TOTL    54211.0   
1  Afghanistan          AFG  Population, total    SP.POP.TOTL  8996351.0   
2       Angola          AGO  Population, total    SP.POP.TOTL  5643182.0   
3      Albania          ALB  Population, total    SP.POP.TOTL  1608800.0   
4      Andorra          AND  Population, total    SP.POP.TOTL    13411.0   

        1961       1962       1963       1964       1965        1966  \
0    55438.0    56225.0    56695.0    57032.0    57360.0     57715.0   
1  9166764.0  9345868.0  9533954.0  9731361.0  9938414.0  10152331.0   
2  5753024.0  5866061.0  5980417.0  6093321.0  6203299.0   6309770.0   
3  1659800.0  1711319.0  1762621.0  1814135.0  1864791.0   1914573.0   
4    14375.0    15370.0    16412.0    17469.0    18549.0     19647.0   

         1967        1968        1969        1970        1971        1972  \
0     58055.0     58386.0     58726.0     59063.0     59440.0     59840.0   
1  10372630.0  10604346.0  10854428.0  11126123.0  11417825.0  11721940.0   
2   6414995.0   6523791.0   6642632.0   6776381.0   6927269.0   7094834.0   
3   1965598.0   2022272.0   2081695.0   2135479.0   2187853.0   2243126.0   
4     20758.0     21890.0     23058.0     24276.0     25559.0     26892.0   

         1973        1974        1975        1976        1977        1978  \
0     60243.0     60528.0     60657.0     60586.0     60366.0     60103.0   
1  12027822.0  12321541.0  12590286.0  12840299.0  13067538.0  13237734.0   
2   7277960.0   7474338.0   7682479.0   7900997.0   8130988.0   8376147.0   
3   2296752.0   2350124.0   2404831.0   2458526.0   2513546.0   2566266.0   
4     28232.0     29520.0     30705.0     31777.0     32771.0     33737.0   

         1979        1980        1981        1982        1983        1984  \
0     59980.0     60096.0     60567.0     61345.0     62201.0     62836.0   
1  13306695.0  13248370.0  13053954.0  12749645.0  12389269.0  12047115.0   
2   8641521.0   8929900.0   9244507.0   9582156.0   9931562.0  10277321.0   
3   2617832.0   2671997.0   2726056.0   2784278.0   2843960.0   2904429.0   
4     34818.0     36067.0     37500.0     39114.0     40867.0     42706.0   

         1985        1986        1987        1988        1989        1990  \
0     63026.0     62644.0     61833.0     61079.0     61032.0     62149.0   
1  11783050.0  11601041.0  11502761.0  11540888.0  11777609.0  12249114.0   
2  10609042.0  10921037.0  11218268.0  11513968.0  11827237.0  12171441.0   
3   2964762.0   3022635.0   3083605.0   3142336.0   3227943.0   3286542.0   
4     44600.0     46517.0     48455.0     50434.0     52448.0     54509.0   

         1991        1992        1993        1994        1995        1996  \
0     64622.0     68235.0     72504.0     76700.0     80324.0     83200.0   
1  12993657.0  13981231.0  15095099.0  16172719.0  17099541.0  17822884.0   
2  12553446.0  12968345.0  13403734.0  13841301.0  14268994.0  14682284.0   
3   3266790.0   3247039.0   3227287.0   3207536.0   3187784.0   3168033.0   
4     56671.0     58888.0     60971.0     62677.0     63850.0     64360.0   

         1997        1998        1999        2000        2001        2002  \
0     85451.0     87277.0     89005.0     90853.0     92898.0     94992.0   
1  18381605.0  18863999.0  19403676.0  20093756.0  20966463.0  21979923.0   
2  15088981.0  15504318.0  15949766.0  16440924.0  16983266.0  17572649.0   
3   3148281.0   3128530.0   3108778.0   3089027.0   3060173.0   3051010.0   
4     64327.0     64142.0     64370.0     65390.0     67341.0     70049.0   

         2003        2004        2005        2006        2007        2008  \
0     97017.0     98737.0    100031.0    100832.0    101220.0    101353.0   
1  23064851.0  24118979.0  25070798.0  25893450.0  26616792.0  27294031.0   
2  18203369.0  18865716.0  19552542.0  20262399.0  20997687.0  21759420.0   
3   3039616.0   3026939.0 

Notice that this population data is the same as the data from the previous exercise. The column order might have changed, but the data is otherwise the same.

# Other Ways to Read in JSON

Besides using pandas to read JSON files, you can use the json library. Run the code cell below to see an example of reading in JSON with the json library. Python treats JSON data like a dictionary.

In [36]:
import json

# read in the JSON file
with open('population_data.json') as f:
    json_data = json.load(f)

# print the first record in the JSON file
print(json_data[0])
print('\n')

# show that JSON data is essentially a dictionary
print(json_data[0]['Country Name'])
print(json_data[0]['Country Code'])

{'Country Name': 'Aruba', 'Country Code': 'ABW', 'Indicator Name': 'Population, total', 'Indicator Code': 'SP.POP.TOTL', '1960': 54211.0, '1961': 55438.0, '1962': 56225.0, '1963': 56695.0, '1964': 57032.0, '1965': 57360.0, '1966': 57715.0, '1967': 58055.0, '1968': 58386.0, '1969': 58726.0, '1970': 59063.0, '1971': 59440.0, '1972': 59840.0, '1973': 60243.0, '1974': 60528.0, '1975': 60657.0, '1976': 60586.0, '1977': 60366.0, '1978': 60103.0, '1979': 59980.0, '1980': 60096.0, '1981': 60567.0, '1982': 61345.0, '1983': 62201.0, '1984': 62836.0, '1985': 63026.0, '1986': 62644.0, '1987': 61833.0, '1988': 61079.0, '1989': 61032.0, '1990': 62149.0, '1991': 64622.0, '1992': 68235.0, '1993': 72504.0, '1994': 76700.0, '1995': 80324.0, '1996': 83200.0, '1997': 85451.0, '1998': 87277.0, '1999': 89005.0, '2000': 90853.0, '2001': 92898.0, '2002': 94992.0, '2003': 97017.0, '2004': 98737.0, '2005': 100031.0, '2006': 100832.0, '2007': 101220.0, '2008': 101353.0, '2009': 101453.0, '2010': 101669.0, '2011'

# Extract XML

Next, you'll work with the same data except now the data is in xml format. Run the next code cell to see what the first fifteen lines of the data file look like.

In [37]:
# Run the code cell to print out the first 15 lines of the xml file
print_lines(15, 'population_data.xml')

<?xml version="1.0" encoding="utf-8"?>

<Root xmlns:wb="http://www.worldbank.org">

  <data>

    <record>

      <field name="Country or Area" key="ABW">Aruba</field>

      <field name="Item" key="SP.POP.TOTL">Population, total</field>

      <field name="Year">1960</field>

      <field name="Value">54211</field>

    </record>

    <record>

      <field name="Country or Area" key="ABW">Aruba</field>

      <field name="Item" key="SP.POP.TOTL">Population, total</field>

      <field name="Year">1961</field>

      <field name="Value">55438</field>

    </record>



XML looks very similar to HTML. XML is formatted with tags having values inside the tags. XML is not as easy to navigate as JSON. Pandas cannot read in XML directly. One reason is that tag names are user defined. Every XML file might have different formatting. You can imagine why XML has fallen out of favor relative to JSON.

### How to read and navigate XML

There is a Python library called BeautifulSoup, which makes reading in and parsing XML data easier. Here is the link to the documentation: [Beautiful Soup Documentation](https://www.crummy.com/software/BeautifulSoup/)

The find() method will find the first place where an xml element occurs. For example using find('record') will return the first record in the xml file:

```xml
<record>
  <field name="Country or Area" key="ABW">Aruba</field>
  <field name="Item" key="SP.POP.TOTL">Population, total</field>
  <field name="Year">1960</field>
  <field name="Value">54211</field>
</record>
```

The find_all() method returns all of the matching tags. So find_all('record') would return all of the elements with the `<record>` tag.

Run the code cells below to get a basic idea of how to navigate XML with BeautifulSoup. To navigate through the xml file, you search for a specific tag using the find() method or find_all() method. 

Below these code cells, there is an exercise for wrangling the XML data.

In [39]:
# import the BeautifulSoup library
from bs4 import BeautifulSoup

# open the population_data.xml file and load into Beautiful Soup
with open("population_data.xml") as fp:
    soup = BeautifulSoup(fp, "lxml") # lxml is the Parser type

In [41]:
# output the first 5 records in the xml file
# this is an example of how to navigate the XML document with BeautifulSoup

i = 0
# use the find_all method to get all record tags in the document
for record in soup.find_all('record'):
    # use the find_all method to get all fields in each record
    i += 1
    for field in record.find_all('field'):
        print(field['name'], ': ' , field.text)
    print()
    if i == 5:
        break

Country or Area :  Aruba
Item :  Population, total
Year :  1960
Value :  54211

Country or Area :  Aruba
Item :  Population, total
Year :  1961
Value :  55438

Country or Area :  Aruba
Item :  Population, total
Year :  1962
Value :  56225

Country or Area :  Aruba
Item :  Population, total
Year :  1963
Value :  56695

Country or Area :  Aruba
Item :  Population, total
Year :  1964
Value :  57032



# XML Exercise (Challenge)

Create a data frame from the xml file. This exercise is somewhat tricky. One solution would be to convert the xml data into dictionaries and then use the dictionaries to create a data frame. 

The dataframe should have the following layout:

| Country or Area | Year | Item | Value |
|----|----|----|----|
| Aruba | 1960 | Population, total | 54211 |
| Aruba | 1961 | Population, total | 55348 |
etc...

Technically, extracting XML, transforming the results, and putting it into a data frame is a full ETL pipeline. This exercise is jumping ahead in terms of what's to come later in the lesson. But it's a good chance to familiarize yourself with XML. 

In [55]:
# TODO: Create a pandas data frame from the XML data.
# HINT: You can use dictionaries to create pandas data frames.
# HINT: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.from_dict.html#pandas.DataFrame.from_dict
# HINT: You can make a dictionary for each column or for each row (See the link above for more information)
# HINT: Modify the code from the previous code cell
with open("population_data.xml") as fp:
    soup = BeautifulSoup(fp, "lxml") # lxml is the Parser type

full_dict = {}
i = 0
for record in soup.find_all('record'):
    record_list = []
    for field in record.find_all('field'):
        record_list.append(field.text)
    full_dict[str(i)] = record_list
    i = i+1

df = pd.DataFrame.from_dict(full_dict, orient='index', columns = ['Country or Area','Item','Year','Value'])     

df = df.pivot(index = 'Country or Area', columns = 'Year',values = 'Value')
df.reset_index(level = 0, inplace = True)

# Conclusion

Like CSV, JSON and XML are ways to format data. If everything is formatted correctly, JSON is especially easy to work with. XML is an older standard and a bit trickier to handle.

As a reminder, there is a solution file for these exercises. You can go to File->Open and then click on 2_extract_exercise.

# Extract Data from SQL Databases

In this workbook, you'll gain experience extracting data from SQL databases. This is an overview of Python tools and assumes you already have experience writing SQL queries.

# Pandas and sqlite3

You can use Pandas to open a SQL database or to run a SQL query against a database. There is more than one way to do this depending on the type of SQL database you are working with: the [sqlite3 library](https://www.sqlite.org/about.html) or the [sqlalchemy library](https://www.sqlalchemy.org/).

In the same folder as this Jupyter notebook, there is a SQLite database file called "population_data.db". SQLite is a database engine meant for single applications. The entire database is contained in one file. You can read more about SQLite [here](https://www.sqlite.org/whentouse.html). 

In this example, the "population_data.db" database contains only one table called "population_data". Run the code in the following cells to see how to use a SQLite database with pandas. (If you're curious how the data was converted from a csv file to a database, go to File->Open and click on create_db.py). You'll find an exercise at the bottom of the Jupyter notebook.

# Demo: SQLite3 and Pandas

In [57]:
import sqlite3
import pandas as pd

# connect to the database
conn = sqlite3.connect('population_data.db')

# run a query
pd.read_sql('SELECT * FROM population_data', conn)

index                                       Country_Name Country_Code  \
0        0                                              Aruba          ABW   
1        1                                        Afghanistan          AFG   
2        2                                             Angola          AGO   
3        3                                            Albania          ALB   
4        4                                            Andorra          AND   
5        5                                         Arab World          ARB   
6        6                               United Arab Emirates          ARE   
7        7                                          Argentina          ARG   
8        8                                            Armenia          ARM   
9        9                                     American Samoa          ASM   
10      10                                Antigua and Barbuda          ATG   
11      11                                          Australia          AUS   
12      12                                            Austria          AUT   
13      13                                         Azerbaijan          AZE   
14      14                                            Burundi          BDI   
15      15                                            Belgium          BEL   
16      16                                              Benin          BEN   
17      17                                       Burkina Faso          BFA   
18      18                                         Bangladesh          BGD   
19      19                                           Bulgaria          BGR   
20      20                                            Bahrain          BHR   
21      21                                       Bahamas, The          BHS   
22      22                             Bosnia and Herzegovina          BIH   
23      23                                            Belarus          BLR   
24      24                                             Belize          BLZ   
25      25                                            Bermuda          BMU   
26      26                                            Bolivia          BOL   
27      27                                             Brazil          BRA   
28      28                                           Barbados          BRB   
29      29                                  Brunei Darussalam          BRN   
30      30                                             Bhutan          BTN   
31      31                                           Botswana          BWA   
32      32                           Central African Republic          CAF   
33      33                                             Canada          CAN   
34      34                     Central Europe and the Baltics          CEB   
35      35                                        Switzerland          CHE   
36      36                                    Channel Islands          CHI   
37      37                                              Chile          CHL   
38      38                                              China          CHN   
39      39                                      Cote d'Ivoire          CIV   
40      40                                           Cameroon          CMR   
41      41                                   Congo, Dem. Rep.          COD   
42      42                                        Congo, Rep.          COG   
43      43                                           Colombia          COL   
44      44                                            Comoros          COM   
45      45                                         Cabo Verde          CPV   
46      46                                         Costa Rica          CRI   
47      47                             Caribbean small states          CSS   
48      48                                               Cuba          CUB   
49      49                                            Curacao          CUW   
50      50                 

In [58]:
pd.read_sql('SELECT "Country_Name", "Country_Code", "1960" FROM population_data', conn)

Country_Name Country_Code  \
0                                                Aruba          ABW   
1                                          Afghanistan          AFG   
2                                               Angola          AGO   
3                                              Albania          ALB   
4                                              Andorra          AND   
5                                           Arab World          ARB   
6                                 United Arab Emirates          ARE   
7                                            Argentina          ARG   
8                                              Armenia          ARM   
9                                       American Samoa          ASM   
10                                 Antigua and Barbuda          ATG   
11                                           Australia          AUS   
12                                             Austria          AUT   
13                                          Azerbaijan          AZE   
14                                             Burundi          BDI   
15                                             Belgium          BEL   
16                                               Benin          BEN   
17                                        Burkina Faso          BFA   
18                                          Bangladesh          BGD   
19                                            Bulgaria          BGR   
20                                             Bahrain          BHR   
21                                        Bahamas, The          BHS   
22                              Bosnia and Herzegovina          BIH   
23                                             Belarus          BLR   
24                                              Belize          BLZ   
25                                             Bermuda          BMU   
26                                             Bolivia          BOL   
27                                              Brazil          BRA   
28                                            Barbados          BRB   
29                                   Brunei Darussalam          BRN   
30                                              Bhutan          BTN   
31                                            Botswana          BWA   
32                            Central African Republic          CAF   
33                                              Canada          CAN   
34                      Central Europe and the Baltics          CEB   
35                                         Switzerland          CHE   
36                                     Channel Islands          CHI   
37                                               Chile          CHL   
38                                               China          CHN   
39                                       Cote d'Ivoire          CIV   
40                                            Cameroon          CMR   
41                                    Congo, Dem. Rep.          COD   
42                                         Congo, Rep.          COG   
43                                            Colombia          COL   
44                                             Comoros          COM   
45                                          Cabo Verde          CPV   
46                                          Costa Rica          CRI   
47                              Caribbean small states          CSS   
48                                                Cuba          CUB   
49                                             Curacao          CUW   
50                                      Cayman Islands          CYM   
51                                              Cyprus          CYP   
52                                      Czech Republic          CZE   
53                                             Germany          DEU   
54                                            Djibouti          DJI   
55                                            Dominica          DM

# Demo: SQLAlchemy and Pandas

If you are working with a different type of database such as MySQL or PostgreSQL, you can use the SQLAlchemy library with pandas. Here are the instructions for connecting to [different types of databases using SQLAlchemy](http://docs.sqlalchemy.org/en/latest/core/engines.html).

Run the code below to see how to connect to the population_data.db database.

In [61]:
import os
print(os.getcwd())

/Users/xuhao3/UDACITY/Data_Engineering


In [63]:
import pandas as pd
from sqlalchemy import create_engine

### 
# create a database engine 
# to find the correct file path, use the python os library:
# import os
# print(os.getcwd())
#
###

engine = create_engine('sqlite:///population_data.db')
pd.read_sql("SELECT * FROM population_data", engine)

index                                       Country_Name Country_Code  \
0        0                                              Aruba          ABW   
1        1                                        Afghanistan          AFG   
2        2                                             Angola          AGO   
3        3                                            Albania          ALB   
4        4                                            Andorra          AND   
5        5                                         Arab World          ARB   
6        6                               United Arab Emirates          ARE   
7        7                                          Argentina          ARG   
8        8                                            Armenia          ARM   
9        9                                     American Samoa          ASM   
10      10                                Antigua and Barbuda          ATG   
11      11                                          Australia          AUS   
12      12                                            Austria          AUT   
13      13                                         Azerbaijan          AZE   
14      14                                            Burundi          BDI   
15      15                                            Belgium          BEL   
16      16                                              Benin          BEN   
17      17                                       Burkina Faso          BFA   
18      18                                         Bangladesh          BGD   
19      19                                           Bulgaria          BGR   
20      20                                            Bahrain          BHR   
21      21                                       Bahamas, The          BHS   
22      22                             Bosnia and Herzegovina          BIH   
23      23                                            Belarus          BLR   
24      24                                             Belize          BLZ   
25      25                                            Bermuda          BMU   
26      26                                            Bolivia          BOL   
27      27                                             Brazil          BRA   
28      28                                           Barbados          BRB   
29      29                                  Brunei Darussalam          BRN   
30      30                                             Bhutan          BTN   
31      31                                           Botswana          BWA   
32      32                           Central African Republic          CAF   
33      33                                             Canada          CAN   
34      34                     Central Europe and the Baltics          CEB   
35      35                                        Switzerland          CHE   
36      36                                    Channel Islands          CHI   
37      37                                              Chile          CHL   
38      38                                              China          CHN   
39      39                                      Cote d'Ivoire          CIV   
40      40                                           Cameroon          CMR   
41      41                                   Congo, Dem. Rep.          COD   
42      42                                        Congo, Rep.          COG   
43      43                                           Colombia          COL   
44      44                                            Comoros          COM   
45      45                                         Cabo Verde          CPV   
46      46                                         Costa Rica          CRI   
47      47                             Caribbean small states          CSS   
48      48                                               Cuba          CUB   
49      49                                            Curacao          CUW   
50      50                 

# Exercise

Connect to the population_data.db SQLite database, and answer the following questions:
1. Write a query that finds the change in population from 1960 to 1961 in Aruba
2. Write a query that finds the population of Belgium and also Luxembourg in 1975. The output should have two rows.

There is a solution if you go to File->Open->3_sql_exercise_solution.ipynb

In [66]:
pd.read_sql('SELECT "1960","1961" FROM population_data where "Country_Name" = "Aruba"', engine)


1960     1961
0  54211.0  55438.0

In [69]:
pd.read_sql('SELECT "Country_Name","1975" FROM population_data where "Country_Name" = "Belgium"OR "Country_Name" ="Luxembourg"', engine)


Country_Name       1975
0      Belgium  9800700.0
1   Luxembourg   358950.0

# APIs

Instead of downloading World Bank data via a csv file, you're going to download the data using the World Bank APIs. The purpose of this exercise is to gain experience with another way of extracting data.

API is an acronym that stands for application programming interface. API’s provide a standardized way for two applications to talk to each other. In this case, the applications communicating with each other are the server application where World Bank stores data and your Jupyter notebook.

If you wanted to pull data directly from the World Bank’s server, you’d have to know what database system the World Bank was using. You’d also need permission to log in directly to the server, which would be a security risk for the World Bank. And if the World Bank ever migrated its data to a new system, you would have to rewrite all of your code again. The API allows you to execute code on the World Bank server without getting direct access.

# Before there were APIs

Before there were APIs, there was web scraping. People would download html directly from a website and then parse the results programatically. This practice is in a legal grey area. One reason that APIs became popular was so that companies could provide data to users and discourage web scraping.

Here are a few articles about the legality of web scraping.

* [QVC Can't Stop Web Scraping](https://www.forbes.com/sites/ericgoldman/2015/03/24/qvc-cant-stop-web-scraping/#120db59b3ca3)
* [Quora - Legality of Web Scraping](https://www.quora.com/What-is-the-legality-of-web-scraping)

All sorts of companies have public facing APIs including Facebook, Twitter, Google and Pinterest. You can pull data from these companies to create your own applications.

In this notebook, you’ll get practice using Python to pull data from the World Bank indicators API.

Here are links to information about the World Bank indicators and projects APIs if you want to learn more:
* [World Bank Indicators API](world bank projects api)
* [World Bank Projects API](http://search.worldbank.org/api/v2/projects)

# Using APIs

In general, you access APIs via the web using a web address. Within the web address, you specify the data that you want. To know how to format the web address, you need to read an API's documentation. Some APIs also require that you send login credentials as part of your request. The World Bank APIs are public and do not require login credentials.

The Python requests library makes working with APIs relatively simple.

# Example Indicators API

Run the code example below to request data from the World Bank Indicators API. According to the documntation, you format your request url like so:

`http://api.worldbank.org/v2/countries/` + list of country abbreviations separated by ; + `/indicators/` + indicator name + `?` + options

where options can include
* per_page - number of records to return per page
* page - which page to return - eg if there are 5000 records and 100 records per page
* date - filter by dates
* format - json or xml
 
 and a few other options that you can read about [here](https://datahelpdesk.worldbank.org/knowledgebase/articles/898581-api-basic-call-structure).


In [77]:
import requests
import pandas as pd

url = 'http://api.worldbank.org/v2/countries/br;cn;us;de/indicators/SP.POP.TOTL/?format=json&per_page=1000'
r = requests.get(url)
r.json()

[{'page': 1,
  'pages': 1,
  'per_page': 1000,
  'total': 240,
  'sourceid': '2',
  'lastupdated': '2019-12-20'},
 [{'indicator': {'id': 'SP.POP.TOTL', 'value': 'Population, total'},
   'country': {'id': 'BR', 'value': 'Brazil'},
   'countryiso3code': 'BRA',
   'date': '2019',
   'value': None,
   'unit': '',
   'obs_status': '',
   'decimal': 0},
  {'indicator': {'id': 'SP.POP.TOTL', 'value': 'Population, total'},
   'country': {'id': 'BR', 'value': 'Brazil'},
   'countryiso3code': 'BRA',
   'date': '2018',
   'value': 209469333,
   'unit': '',
   'obs_status': '',
   'decimal': 0},
  {'indicator': {'id': 'SP.POP.TOTL', 'value': 'Population, total'},
   'country': {'id': 'BR', 'value': 'Brazil'},
   'countryiso3code': 'BRA',
   'date': '2017',
   'value': 207833831,
   'unit': '',
   'obs_status': '',
   'decimal': 0},
  {'indicator': {'id': 'SP.POP.TOTL', 'value': 'Population, total'},
   'country': {'id': 'BR', 'value': 'Brazil'},
   'countryiso3code': 'BRA',
   'date': '2016',
   '

In [78]:
###
# Run this cell that converts the json into a dataframe
# Note that you do not need the pd.read_json() method because this is not a file or a string containing json 
##

pd.DataFrame(r.json()[1])

indicator  \
0    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
1    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
2    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
3    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
4    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
5    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
6    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
7    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
8    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
9    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
10   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
11   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
12   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
13   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
14   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
15   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
16   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
17   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
18   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
19   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
20   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
21   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
22   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
23   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
24   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
25   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
26   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
27   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
28   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
29   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
30   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
31   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
32   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
33   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
34   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
35   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
36   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
37   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
38   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
39   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
40   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
41   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
42   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
43   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
44   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
45   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
46   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
47   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
48   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
49   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
50   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
51   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
52   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
53   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
54   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
55   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
56   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
57   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
58   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
59   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
60   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
61   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
62   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
63   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
64   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
65   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
66   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
67   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
68   {'id': 'SP.POP.TOTL', 'value': 'Popula

This json data isn't quite ready for a pandas data frame. Notice that the json response is a list with two entries. The first entry is 
```
{''page': 1,
  'pages': 1,
  'per_page': 1000,
  'total': 240,
  'sourceid': '2',
  'lastupdated': '2019-12-20'}
```

That first entry is meta data about the results. For example, it says that there is one page returned with 232 results. 

The second entry is another list containing the data. This data would need some cleaning to be used in a pandas data frame. That would happen later in the transformation step of an ETL pipeline. Run the cell below to read the results into a dataframe and see what happens.

In [79]:
###
# Run this cell that converts the json into a dataframe
# Note that you do not need the pd.read_json() method because this is not a file or a string containing json 
##

pd.DataFrame(r.json()[1])

indicator  \
0    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
1    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
2    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
3    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
4    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
5    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
6    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
7    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
8    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
9    {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
10   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
11   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
12   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
13   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
14   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
15   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
16   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
17   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
18   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
19   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
20   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
21   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
22   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
23   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
24   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
25   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
26   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
27   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
28   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
29   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
30   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
31   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
32   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
33   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
34   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
35   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
36   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
37   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
38   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
39   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
40   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
41   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
42   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
43   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
44   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
45   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
46   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
47   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
48   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
49   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
50   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
51   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
52   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
53   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
54   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
55   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
56   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
57   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
58   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
59   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
60   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
61   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
62   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
63   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
64   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
65   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
66   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
67   {'id': 'SP.POP.TOTL', 'value': 'Population, to...   
68   {'id': 'SP.POP.TOTL', 'value': 'Popula

There are some issues with this dataframe. The country and indicator variables don't look particularly useful in their current form. Again, dealing with those issues would come in the transformation phase of a pipeline, which comes later in the lesson.

# Exercise Indicators API

Use the Indicators API to request rural population data for Switzerland in the years 1995 through 2001. Here are a few helpful resources:
* [documentation included how to filter by year](https://datahelpdesk.worldbank.org/knowledgebase/articles/898581-api-basic-call-structure)
* [2-character iso country codes](https://www.nationsonline.org/oneworld/country_code_list.htm)
* [search box for World Bank indicators](https://data.worldbank.org)

To find the indicator code, first search for the indicator here: https://data.worldbank.org
Click on the indicator name. The indicator code is in the url. For example, the indicator code for total population is SP.POP.TOTL, which you can see in the link [https://data.worldbank.org/indicator/SP.RUR.TOTL](https://data.worldbank.org/indicator/SP.RUR.TOTL).

In [82]:
# TODO: get the url ready
url = 'http://api.worldbank.org/v2/countries/ch/indicators/SP.RUR.TOTL/?format=json&per_page=1000&date=1995:2001'

# TODO: send the request
r = requests.get(url)

# TODO: output the json using the json method like in the previous example
pd.DataFrame(r.json()[1])


indicator  \
0  {'id': 'SP.RUR.TOTL', 'value': 'Rural populati...   
1  {'id': 'SP.RUR.TOTL', 'value': 'Rural populati...   
2  {'id': 'SP.RUR.TOTL', 'value': 'Rural populati...   
3  {'id': 'SP.RUR.TOTL', 'value': 'Rural populati...   
4  {'id': 'SP.RUR.TOTL', 'value': 'Rural populati...   
5  {'id': 'SP.RUR.TOTL', 'value': 'Rural populati...   
6  {'id': 'SP.RUR.TOTL', 'value': 'Rural populati...   

                                country countryiso3code  date    value unit  \
0  {'id': 'CH', 'value': 'Switzerland'}             CHE  2001  1924949        
1  {'id': 'CH', 'value': 'Switzerland'}             CHE  2000  1912232        
2  {'id': 'CH', 'value': 'Switzerland'}             CHE  1999  1897587        
3  {'id': 'CH', 'value': 'Switzerland'}             CHE  1998  1884719        
4  {'id': 'CH', 'value': 'Switzerland'}             CHE  1997  1875299        
5  {'id': 'CH', 'value': 'Switzerland'}             CHE  1996  1866968        
6  {'id': 'CH', 'value': 'Switzerland'}             CHE  1995  1854939        

  obs_status  decimal  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  
5                   0  
6                   0